man/addusr.1
.TH ADDUSR 1 "29 Nov 2007" "Project Athena"
\" RCSID: $HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/man/addusr.1 $ $Id: addusr.1 3956 2010-01-05 20:56:56Z zacheiss $
.SH NAME
addusr \- add and/or register users in Moira
.SH SYNOPSIS
.B addusr [options]
.SH DESCRIPTION
.I Addusr
is a tool for adding users in bulk to the moira database and
registering them for accounts.  It reads the users from a file (or
stdin) rather than the menu-oriented regular moira client.

The users are entered in one of two formats, depending on whether or
not they will be registered.  When users are only being added to the
database, the format is:
.nf
  lastname, firstname middlename idnumber
.fi
The lastname may have spaces in it, as it will continue up until the
first comma on the line.  The firstname also may have spaces, as it
will take up everything but the last two tokens on the line.  If only
3 tokens are on the line, the middlename will be assumed to be the
empty string.  The middlename cannot have spaces in it.  The idnumber
also cannot have spaces, although it may have hyphens in it.  The
capitalization of the names is not important as it will be corrected
before entering into the database, unless the \fB-n\fR flag has been
specified.  If the operation specified for
.B addusr
includes registering users, then there is one additional token on
the end of the line:
.nf
  lastname, firstname middlename idnumber loginname
.fi
.SH OPTIONS
.IP \fB-file\ \fIfilename\ \fRor\ \fB-f\ \fIfilename\fR:
This specifies the file that will be read for the list of users to
operate on.  If not specifed, standard input will be used.  The file
format is described above.
.IP \fB-class\ \fIclass\ \fRor\ \fB-c\ \fIclass\fR:
This specifies the account class that will be used for all users added
in this invocation.  If not specified, \fITEMP\fR will be used.
.IP \fB-comment\ \fI"comment"\ \fRor\ \fB-C\ \fI"comment"\fR:
This specifies the comment that will be stored for all users added
in this invocation.  If not specified, the comment field will be left
blank.  Note that the double-quotes are not required by \fBaddusr\fR,
but will be parsed out by the shell and are only needed to specify
comments with spaces in them.
.IP \fB-status\ \fIstatus\ \fRor\ \fB-s\ \fIstatus\fR:
This specifies the account status that will be used for all users added
in this invocation.  If not specified, \fI0 (registerable)\fR will be used.
.IP \fB-shell\ \fIshell\ \fRor\ \fB-h\ \fIshell\fR:
This specifies the login shell that will be stored for all users added
in this invocation.  If not specified, a default will be used.  At MIT
Athena the default is \fI/bin/athena/bash\fR, elsewhere it is likely to be 
\fI/bin/bash\fR.
.IP \fB-winshell\ \fIshell\ \fRor\ \fB-w\ \fIshell\fR:
This specifies the Windows console shell that will be used for all users
added in this invocation.  If not specified, a default value of
\fIcmd\fR will be used.
.IP \fB-sponsor\ \fIsponsor\ \fRor\ \fB-sp\ \fIsponsor\fR:
This specifies the USER, LIST, or KERBEROS that will be listed as the
sponsor for all users added in this invocation.  This is intended for
use with accounts in the GUEST and VOUCH classes.
.IP \fB-expiration\ \fIdate\ \fRor\ \fB-e\ \fIdate\fR:
This specified the expiration date that will be recorded for all users
added in this invocation.  This is intended for use with accounts in the
GUEST or VOUCH classes. 
.IP \fB-register\ \fRor\ \fB-R\fR:
This instructs \fBaddusr\fR to both add the users to the database and
then to immediately register them. The users will be left
"half-registered", meaning that they have everything but a Kerberos
principal.
.IP \fB-reg_only\ \fRor\ \fB-r\fR:
This instructs \fBaddusr\fR to (half-) register the users without
attempting to add them to the database. They must already exist in the
database.
.IP \fB-verbose\ \fRor\ \fB-v\fR:
This causes \fBaddusr\fR to print messages about each operation as it
performs it.
.IP \fB-nodupcheck\ \fRor\ \fB-d\fR:
Do not attempt to check the database for duplicates before adding
users.  Normally \fBaddusr\fR will check the database for users with
the same name.  If it finds any who match in both name and ID number,
the user will not be added.  If it finds a name match with a different
ID number, it will print a warning and add the user anyway.
.IP \fB-nocaps\ \fRor\ \fB-n\fR:
Do not attempt to normalize the capitalization of first, middle, and
last names before adding users.  Normally, \fBaddusr\fR will update the
capitalization of names based on a set of heuristics.  This flag can be
used to suppress this behavior if it is inappropriate for certain names.
.IP \fB-database\ \fIhost:port\ \fRor\ \fB-db\ \fIhost:port\fR:
Use the specified host and port to contact the Moira server instead of
the default server.  Both may be symbolic names or numbers.  If the
port is left off, the default Moira server port will be assumed.  The
server chosen will be the one specified on the command line, specified
in the MOIRASERVER environment variable, the hesiod "moira" sloc entry,
or the compiled in default, in that order or preference.
.IP \fB-secure\ \fRor\ \fB-6\fR:
This specifies that users added in this invocation will need a secure
"Account Coupon" containing additional verification information beyond
name and ID number to be allowed to register.

.SH AUTHOR
Mark Rosenstein
.SH SEE ALSO
moira(1), userreg(1)


debian/changelog
debathena-moira (4.0.0+svn20100225-0debathena2) UNRELEASED; urgency=low

  * There is no "lib" section - it's "libs".

 -- Evan Broder <broder@mit.edu>  Fri, 26 Feb 2010 12:50:19 -0500

debathena-moira (4.0.0+svn20100225-0debathena1) unstable; urgency=low

  * New SVN snapshot.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 15:54:26 -0500

debathena-moira (4.0.0+svn20100223-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Drop patch install-headers, which was incorporated upstream.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 02:13:44 -0500

debathena-moira (4.0.0+svn20100104-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Build and install libmrclient as a shared library.
  * Deal with rename of /etc{/athena => }/moira.conf
  * Patch manpages to deal with our installation paths.

 -- Evan Broder <broder@mit.edu>  Mon, 04 Jan 2010 22:05:10 -0500

debathena-moira (4.0.0+cvs20091116-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop patch fix-update-ticket, which was incorporated upstream.
  * Add a missing definition of ${prefix} in include/Makefile.in so that
    the moira.h header actually gets installed in
    /usr/include/moira. (Trac: #445)

 -- Evan Broder <broder@mit.edu>  Thu, 19 Nov 2009 10:53:40 -0500

debathena-moira (4.0.0+cvs20090922-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Add patch to fix builds without krb4.

 -- Evan Broder <broder@mit.edu>  Tue, 22 Sep 2009 00:35:12 -0400

debathena-moira (4.0.0+cvs20090728-0debathena3) unstable; urgency=low

  * Add a missing conflicts to make upgrades run more smoothly.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 18:28:33 -0700

debathena-moira (4.0.0+cvs20090728-0debathena2) unstable; urgency=low

  * Include the headers in debathena-libmoira-dev.
  * Install update_test in moira-clients instead of moira-update-server.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 17:54:34 -0700

debathena-moira (4.0.0+cvs20090728-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Fix zephyr dependencies to actually pull in Kerberized libraries; drop
    build-dep on Kerberized library.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 13:43:28 -0400

debathena-moira (4.0.0+cvs20090621-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop krb4less-fix; it was incorporated upstream.
  * Universally stop building Moira with krb4 support.

 -- Evan Broder <broder@mit.edu>  Fri, 26 Jun 2009 00:51:28 -0400

debathena-moira (4.0.0+cvs20090618-0debathena1) unstable; urgency=low

  * New upstream release
  * Build and install libmoira as a shared library. (Trac: #70)
  * Drop most of the krb4 patch - it's been incorporated upstream.
  * Install the Moira development headers by patching the relevant
    Makefiles, instead of in the debian/rules file.

 -- Evan Broder <broder@mit.edu>  Sat, 20 Jun 2009 14:45:32 -0400

debathena-moira (4.0.0+cvs20090409-0debathena1) unstable; urgency=low

  * New CVS snapshot (Trac: #195)
  * Drop patches that have been incorporated upstream.
  * Add patch to allow building on systems without krb4. (Trac: #198)

 -- Evan Broder <broder@mit.edu>  Sun, 12 Apr 2009 14:03:22 -0400

debathena-moira (4.0.0+cvs20080830-0debathena2) unstable; urgency=low

  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 15:32:35 -0500

debathena-moira (4.0.0+cvs20080830-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Add a debathena-moira-update-server package that installs
    update_server

 -- Evan Broder <broder@mit.edu>  Fri, 01 Aug 2008 10:54:26 -0400

debathena-moira (4.0.0+cvs20080701-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Move Moira header files from /usr/share/include to /usr/include/moira

 -- Evan Broder <broder@mit.edu>  Tue, 01 Jul 2008 20:38:13 -0400

debathena-moira (4.0.0+cvs20080304-0debathena1) unstable; urgency=low

  * New CVS snapshot to fix chsh.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 04 Mar 2008 15:50:31 -0500

debathena-moira (4.0.0+cvs20080301-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Remove mrtest segfault patch from last release (fixed in upstream)

 -- Tim Abbott <tabbott@mit.edu>  Sat,  1 Mar 2008 18:51:40 -0500

debathena-moira (4.0.0+cvs20070817-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Patch an mrtest segfault with too many command line arguments.

 -- Anders Kaseorg <andersk@mit.edu>  Fri, 24 Aug 2007 19:53:49 -0400

debathena-moira (4.0.0+cvs20070129-0debathena3) unstable; urgency=low

  * Move chsh.debathena to chsh.moira and chfn.debathena to chfn.moira.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 06 Aug 2007 03:17:10 -0400

debathena-moira (4.0.0+cvs20070129-0debathena2) unstable; urgency=low

  * Add missing build dependencies to libmoira-dev.
  * Section: debathena/net.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 17 Jul 2007 06:34:07 -0400

debathena-moira (4.0.0+cvs20070129-0debathena1) unstable; urgency=low

  * Initial release.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 14 Feb 2007 00:53:24 -0500


clients/addusr/addusr.c
/* $Id: addusr.c 3987 2010-02-25 19:53:43Z zacheiss $
 *
 * Program to add users en masse to the moira database
 *
 * by Mark Rosenstein, July 1992.
 *
 * Copyright (C) 1992-1998 by the Massachusetts Institute of Technology.
 * For copying and distribution information, please see the file
 * <mit-copyright.h>.
 */

#include <mit-copyright.h>
#include <moira.h>
#include <moira_site.h>
#include <mrclient.h>

#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>

RCSID("$HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/clients/addusr/addusr.c $ $Id: addusr.c 3987 2010-02-25 19:53:43Z zacheiss $");

struct owner_type {
  int type;
  char *name;
};

#define M_ANY		0
#define M_USER		1
#define M_LIST		2
#define M_KERBEROS	3
#define M_NONE		4

#ifdef ATHENA
#define DEFAULT_SHELL "/bin/athena/bash"
#else
#define DEFAULT_SHELL "/bin/bash"
#endif

#define DEFAULT_WINCONSOLESHELL "cmd"
#define DEFAULT_WINHOMEDIR "[DFS]"
#define DEFAULT_WINPROFILEDIR "[DFS]"

/* flags from command line */
char *class, *comment, *status_str, *shell, *winconsoleshell, *filename;
char *expiration;
int reg_only, reg, verbose, nodupcheck, securereg, nocaps;
struct owner_type *sponsor;

/* argument parsing macro */
#define argis(a, b) (!strcmp(*arg + 1, a) || !strcmp(*arg + 1, b))

char *whoami;
int duplicate, errors;

void usage(char **argv);
int usercheck(int argc, char **argv, void *qargv);
int get_uid(int argc, char **argv, void *qargv);
struct owner_type *parse_member(char *s);

int main(int argc, char **argv)
{
  int status, lineno;
  char **arg = argv, *qargv[U_END];
  char *p, *first, *middle, *last, *id, *login, *server;
  char buf[BUFSIZ];
  FILE *input;

  /* clear all flags & lists */
  reg_only = reg = verbose = lineno = nodupcheck = errors = securereg = nocaps = 0;
  sponsor = NULL;
  server = NULL;
  filename = "-";
  shell = DEFAULT_SHELL;
  winconsoleshell = DEFAULT_WINCONSOLESHELL;
  class = "TEMP";
  comment = "";
  expiration = "";
  status_str = "0";

  whoami = argv[0];

  /* parse args */
  while (++arg - argv < argc)
    {
      if  (**arg == '-')
	{
	  if (argis("c", "class"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  class = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("C", "comment"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  comment = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("s", "status"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  status_str = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("h", "shell"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  shell = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("w", "winshell"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  winconsoleshell = *arg;
		}
	      else 
		usage(argv);
	    }
	  else if (argis("sp", "sponsor"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  sponsor = parse_member(*arg);
		}
	      else
		usage(argv);
	    }
	  else if (argis("e", "expiration"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  expiration = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("6", "secure"))
	    securereg++;
	  else if (argis("r", "reg_only"))
	    reg_only++;
	  else if (argis("R", "register"))
	    reg++;
	  else if (argis("f", "file"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  filename = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("v", "verbose"))
	    verbose++;
	  else if (argis("d", "nodupcheck"))
	    nodupcheck++;
	  else if (argis("n", "nocaps"))
	    nocaps++;
	  else if (argis("S", "server") || argis("db", "database"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  server = *arg;
		}
	      else
		usage(argv);
	    }
	  else
	    usage(argv);
	}
      else
	usage(argv);
    }

  if (!strcmp(filename, "-"))
    input = stdin;
  else
    {
      input = fopen(filename, "r");
      if (!input)
	{
	  com_err(whoami, errno, "opening input file %s", filename);
	  exit(2);
	}
    }

  /* fire up Moira */
  if (mrcl_connect(server, "addusr", 14, 1) != MRCL_SUCCESS)
    exit(2);

  qargv[U_NAME] = UNIQUE_LOGIN;
  qargv[U_UID] = UNIQUE_UID;
  qargv[U_SHELL] = shell;
  qargv[U_WINCONSOLESHELL] = winconsoleshell;
  qargv[U_WINHOMEDIR] = DEFAULT_WINHOMEDIR;
  qargv[U_WINPROFILEDIR] = DEFAULT_WINPROFILEDIR;
  qargv[U_EXPIRATION] = expiration;
  qargv[U_ALT_EMAIL] = "";
  qargv[U_ALT_PHONE] = "";
  qargv[U_STATE] = status_str;
  qargv[U_CLASS] = class;
  qargv[U_COMMENT] = comment;
  qargv[U_SIGNATURE] = "";
  qargv[U_SECURE] = securereg ? "1" : "0";
  
  while (fgets(buf, BUFSIZ, input))
    {
      /* throw away terminating newline */
      p = &buf[strlen(buf) - 1];
      if (*p == '\n')
	*p = '\0';
      lineno++;
      if (strlen(buf) == 0)
	continue;
      /* Last name is first thing on line */
      last = buf;
      /* First name follows a comma */
      p = strchr(last, ',');
      if (!p)
	{
	  com_err(whoami, MR_BAD_CHAR, "Missing comma on line %d", lineno);
	  errors++;
	  continue;
	}
      *p++ = '\0';
      first = strtrim(p);
      while (*p)		/* find end-of-line */
	p++;
      if (reg_only || reg)
	{
	  while (!isspace(*p))
	    p--;
	  if (p <= first)
	    {
	      com_err(whoami, 0, "Missing login on line %d", lineno);
	      errors++;
	      continue;
	    }
	  login = strtrim(&p[1]);
	  *p-- = '\0';
	}
      else
	{
	  /* ID is last thing on line */
	  p--;
	}
      while (isspace(*p))
	p--;
      while (!isspace(*p))
	p--;
      if (p <= first)
	{
	  com_err(whoami, 0, "Missing ID on line %d", lineno);
	  errors++;
	  continue;
	}
      id = &p[1];
      *p-- = '\0';
      /* If something between first name & ID, it's middle name */
      while (isspace(*p))
	p--;
      while (!isspace(*p))
	p--;
      if (p <= first)
	middle = "";
      else
	{
	  middle = &p[1];
	  *p = '\0';
	}
      qargv[U_FIRST] = strtrim(first);
      qargv[U_MIDDLE] = strtrim(middle);
      qargv[U_LAST] = strtrim(last);
      qargv[U_MITID] = strtrim(id);
      if (!nocaps)
	{
	  FixCase(qargv[U_FIRST]);
	  FixCase(qargv[U_MIDDLE]);
	  FixCase(qargv[U_LAST]);
	}
      RemoveHyphens(qargv[U_MITID]);
      if (!reg_only)
	{
	  if (!nodupcheck)
	    {
	      char *dargv[2];

	      dargv[0] = qargv[U_FIRST];
	      dargv[1] = qargv[U_LAST];
	      duplicate = 0;
	      status = mr_query("get_user_account_by_name", 2, dargv,
				usercheck, qargv);
	      if (status && status != MR_NO_MATCH)
		{
		  com_err(whoami, status,
			  "checking to see if user %s %s already exists",
			  qargv[U_FIRST], qargv[U_LAST]);
		  com_err(whoami, 0, "NOT ADDING USER");
		  errors++;
		  continue;
		}
	      if (duplicate > 0)
		{
		  com_err(whoami, MR_EXISTS, "user %s %s already exists",
			  qargv[U_FIRST], qargv[U_LAST]);
		  com_err(whoami, 0, "NOT ADDING USER");
		  errors++;
		  continue;
		}
	      else if (duplicate < 0)
		{
		  com_err(whoami, MR_EXISTS,
			  "user %s %s already exists with different ID number",
			  qargv[U_FIRST], qargv[U_LAST]);
		  com_err(whoami, 0, "ADDING user anyway");
		}
	    }

	  if (sponsor)
	    {
	      qargv[U_SPONSOR_NAME] = sponsor->name;
	      switch (sponsor->type)
		{
		case M_ANY:
		case M_USER:
		  qargv[U_SPONSOR_TYPE] = "USER";
		  status = mr_query("add_user_account", 20, qargv, NULL, NULL);
		  if (sponsor->type != M_ANY || status != MR_USER)
		    break;
		  
		case M_LIST:
		  qargv[U_SPONSOR_TYPE] = "LIST";
		  status = mr_query("add_user_account", 20, qargv, NULL, NULL);
		  break;
		  
		case M_KERBEROS:
		  qargv[U_SPONSOR_TYPE] = "KERBEROS";
		  status = mrcl_validate_kerberos_member(qargv[U_SPONSOR_NAME],
							 &qargv[U_SPONSOR_NAME]);
		  if (mrcl_get_message())
		    mrcl_com_err(whoami);
		  if (status == MRCL_REJECT)
		exit(1);
		  status = mr_query("add_user_account", 20, qargv, NULL, NULL);
		  break;
		  
		case M_NONE:
		  qargv[U_SPONSOR_TYPE] = "NONE";
		  status = mr_query("add_user_account", 20, qargv, NULL, NULL);
		  break;
		}
	    }
	  else
	    {
	      qargv[U_SPONSOR_TYPE] = "NONE";
	      qargv[U_SPONSOR_NAME] = "NONE";
	  
	      status = mr_query("add_user_account", 20, qargv, NULL, NULL);
	    }
	  
	  if (status)
	    {
	      com_err(whoami, status, "adding user %s %s", first, last);
	      errors++;
	    }
	  else if (verbose)
	    {
	      printf("Added user %s %s %s (%s)\n", qargv[U_FIRST],
		     qargv[U_MIDDLE], qargv[U_LAST], qargv[U_MITID]);
	    }
	}
      if (reg || reg_only)
	{
	  char *gargv[2], *rargv[3], uid[10];

	  uid[0] = '\0';
	  gargv[0] = qargv[U_FIRST];
	  gargv[1] = qargv[U_LAST];
	  status = mr_query("get_user_account_by_name", 2, gargv,
			    get_uid, &uid);
	  if (status)
	    {
	      com_err(whoami, status, "while looking up uid for %s %s",
		      qargv[U_FIRST], qargv[U_LAST]);
	      errors++;
	      continue;
	    }

	  rargv[0] = uid;
	  rargv[1] = login;
	  rargv[2] = "EXCHANGE";

	  status = mr_query("register_user", 3, rargv, NULL, NULL);
	  if (status)
	    {
	      com_err(whoami, status, "while registering (login) %s %s",
		      qargv[U_FIRST], qargv[U_LAST]);
	      errors++;
	      continue;
	    }
	  else if (verbose)
	    {
	      printf("Registered user %s %s as %s\n", qargv[U_FIRST],
		     qargv[U_LAST], login);
	    }
	}
    }

  exit(errors);
}


void usage(char **argv)
{
  fprintf(stderr, "Usage: %s [options]\n", argv[0]);
  fprintf(stderr, "Options are\n");
  fprintf(stderr, "   -f | -file filename (default STDIN)\n");
  fprintf(stderr, "   -c | -class class (default TEMP)\n");
  fprintf(stderr, "   -C | -comment \"comment\" (default \"\")\n");
  fprintf(stderr, "   -s | -status status (default 0)\n");
  fprintf(stderr, "   -h | -shell shell (default %s)\n", DEFAULT_SHELL);
  fprintf(stderr, "   -w | -winshell windows console shell (default %s)\n",
	  DEFAULT_WINCONSOLESHELL);
  fprintf(stderr, "   -sp | -sponsor sponsor (default NONE)\n");
  fprintf(stderr, "   -e | -expiration \"expiration date\" (default \"\")\n");
  fprintf(stderr, "   -r | -reg_only\n");
  fprintf(stderr, "   -R | -register (and add to database)\n");
  fprintf(stderr, "   -v | -verbose\n");
  fprintf(stderr, "   -d | -nodupcheck (don't check for duplicates)\n");
  fprintf(stderr, "   -n | -nocaps (don't fix capitalization of names)\n");
  fprintf(stderr, "   -db | -database host:port\n");
  exit(1);
}


/* query callback routine to check for duplicate users */

int usercheck(int argc, char **argv, void *qargv)
{
  if (!strcmp(argv[U_MITID], ((char **)qargv)[U_MITID]))
    duplicate++;
  else
    duplicate--;

  return MR_CONT;
}

/* query callback to get uid of a just-added account */
int get_uid(int argc, char **argv, void *uidv)
{
  char *uid = uidv;

  if (uid[0] == '\0')
    strcpy(uid, argv[U_UID]);
  else
    {
      if (!strcmp(argv[U_MODWITH], "addusr"))
	strcpy(uid, argv[U_UID]);
    }

  return MR_CONT;
}

/* Parse a line of input, fetching a member.  NULL is returned if a member
 * is not found.  ';' is a comment character.
 */
struct owner_type *parse_member(char *s)
{
  struct owner_type *m;
  char *p, *lastchar;

  while (*s && isspace(*s))
    s++;
  lastchar = p = s;
  while (*p && *p != '\n' && *p != ';')
    {
      if (isprint(*p) && !isspace(*p))
	lastchar = p++;
      else
	p++;
    }
  lastchar++;
  *lastchar = '\0';
  if (p == s || strlen(s) == 0)
    return NULL;

  if (!(m = malloc(sizeof(struct owner_type))))
    return NULL;

  if ((p = strchr(s, ':')))
    {
      *p = '\0';
      m->name = ++p;
      if (!strcasecmp("user", s))
	m->type = M_USER;
      else if (!strcasecmp("list", s))
	m->type = M_LIST;
      else if (!strcasecmp("kerberos", s))
	m->type = M_KERBEROS;
      else if (!strcasecmp("none", s))
	m->type = M_NONE;
      else
	{
	  m->type = M_ANY;
	  *(--p) = ':';
	  m->name = s;
	}
      m->name = strdup(m->name);
    }
  else
    {
      m->name = strdup(s);
      m->type = strcasecmp(s, "none") ? M_ANY : M_NONE;
    }
  return m;
}
